### ISSUE TITLE
local unbound variable not detected

### ISSUE THREAD

**Issue by jaysoffian** [koalaman/shellcheck#1611]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or "online"): online and 0.6.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

set -u

fn1() {
        local foo=foo
        echo "$foo"
}

fn2() {
        echo "$foox"
}

fn1
fn2
```

#### Here's what shellcheck currently says:

Nothing. :-(

#### Here's what I wanted or expected to see:

```
Line 11:
        echo "$foo"
              ^-- SC2154: foo is referenced but not assigned.
```

**Comment by matthewpersico** [comment]
Huh? You just assigned the string “foo” to the variable $foo in the previous line; what am I missing?

**Comment by jaysoffian** [comment]
@matthewpersico I updated the description to be clearer. `foo` is local to `fn1` so ShellCheck should warn about the reference to `$foo` in `fn2`.

**Comment by matthewpersico** [comment]
Duh. Seems I can’t count to 11. 🤦🏻‍♂️ Thanks.

**Comment by SebastianSimon** [comment]
This is the same issue as ShellCheck failing to warn about the unused variable `foo` (in function `y`) in this script, isn’t it?

```sh
#!/bin/bash

x(){
  local foo='foo'

  echo "${foo}"
}

y(){
  local foo='foo' # This is unused! Why doesn’t ShellCheck warn about it?

  echo "no foo"
}

x
y
```

**Comment by nexushoratio** [comment]
```shell
#!/bin/bash

set -u

a() {
  local x=0
}

b() {
  echo "${x}"
}
```

It has both an unused variable and and undefined variable.

But, what about this:

```shell
#!/bin/bash

set -u

a() {
  local x=2
  c
}

b() {
  echo "${x}"
}

c() {
  b
}

a
c
```

`b` can see `x` when `a` calls `c`.
But when `c` is called directly, no `x` is present.

---

In a larger script, perhaps `b` gets called from a large variety of paths.  So each one would have to be investigated to see if `x` is present in each path.  And chasing each path would get very resource intensive.

**Comment by wileyhy** [comment]
Additionally, Shell really wasn't intended for a Functional Programming
paradigm.

Functions in Bash are useful for eliminating repetitive code.

**Comment by markjreed** [comment]
It seems that local variables don't trigger SC2154 even without the assignment.

```
foo() {
    local myvar
    echo "$myvar"
}
```

the `local` declaration doesn't initialize the variable; it's still unbound, and the echo will still trigger an error if run under `set -u`.  I don't see the logic behind excluding such variables from SC2154.

**Comment by wileyhy** [comment]
Please forgive me if this is review, and also, please correct me if I am
wrong.

In shell, variables can be in one of three states: "unset," "declared," or
"assigned."

~ $ declare -p foo
bash: declare: foo: not found
~ $ declare foo # "declared"
~ $ declare -p foo
declare -- foo
~ $ foo=51 # "assigned"
~ $ declare -p foo
declare -- foo="51"
~ $ foo="" # still "assigned"
~ $ declare -p foo
declare -- foo=""
~ $ unset foo # "unset"
~ $ declare -p foo
bash: declare: foo: not found
~ $

In Bash, when a variable is "unset," it doesn't exist in the program's
memory. Attribute(s) can be "declared," and values can be "assigned." When
it's been "declared," then a record of some parameter exists and has an
identifier, but a "declared" variable doesn't necessarily have any
attributes (beyond existing) or any values connected to it.

When "${foo}" has the value '51' in the example above, it doesn't have any
other attributes associated with it (again, beyond existing). Similarly, a
variable can have an attribute, as in this next example with 'readonly',
and can at the same time not have an assigned value.

~ $ declare -p foo
bash: declare: foo: not found
~ $ readonly foo
~ $ declare -p foo
declare -r foo
~ $ foo=""
bash: foo: readonly variable
~ $ declare -p foo
declare -r foo

Notably, the NUL byte is a valid value. These assignment statements are
synonymous:

~ $ set -x; a= b="" c=$'\0'; declare -p a b c
+ a=
+ b=
+ c=
+ declare -p a b c
declare -- a=""
declare -- b=""
declare -- c=""

Typically, variables have both attributes and values, but variables which
have been destroyed with the `unset` built-in have neither of these
(including existing).

In Bash, there are "declaration commands," "`alias`, `declare`, `typeset`,
`export`, `readonly`, and `local` builtin commands," as well as the `unset`
built-in command. These built-ins are used for creating and destroying
variables. There are also "Assignment statements," of which there are a
variety of valid syntax, including `x=y`.

'Nounset' tests for whether a variable has been assigned a value.

~ $ set -u
~ $ unset foo
~ $ echo "$foo"
bash: foo: unbound variable
~ $ declare foo
~ $ echo "$foo"
bash: foo: unbound variable
~ $ declare -p foo
declare -- foo
~ $ export foo
~ $ echo "$foo"
bash: foo: unbound variable
~ $ declare -p foo
declare -x foo
~ $ foo=51
~ $ echo "$foo"
51
~ $

Under errexit and nounset, `local myvar` declares a variable and returns
with an exit code of 0: the variable exists. `echo "$myvar" fails with an
"unbound" message because myvar hadn't been assigned a value yet.

"Initializing" isn't a term used in Bash's documentation, IIRC, but I
understand the idea. The question is whether one maps it to "declaration"
or "assignment." At declaration, the first step in a variable's life has
taken place, so some initialization has occurred, but 'nounset' still tests
for assignment.

I might be wrong, but my understanding is that in historical versions of
Bourne shell, manual declaration of variables was necessary, and that
automatic declaration is a modernization since then. So, in an earlier
version, if someone wrote `echo "$foo"` it would have been assumed that
there was a `foo=bar export foo` statement of some kind or other somewhere
earlier in the script. So the word "unset" in the name "nounset" wouldn't
have logically refered to a declaration command: Why would a shell
programmer try to expand a variable that they hadn't manually declared
yet?  My guess is that "unset" originally meant "unassigned."

Hope that helps,
Wiley
